sound/soc/sof/core.c

```diff
@@ -218,20 +240,13 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	/* hereafter all FW boot flows are for PM reasons */
 	sdev->first_boot = false;
 
-	/* now register audio DSP platform driver and dai */
-	ret = devm_snd_soc_register_component(sdev->dev,
-					      &sdev->sof_audio_data->plat_drv,
-					      sof_ops(sdev)->drv,
-					      sof_ops(sdev)->num_drv);
-	if (ret < 0) {
-		dev_err(sdev->dev,
-			"error: failed to register DSP DAI driver %d\n", ret);
-		goto fw_trace_err;
-	}
+	INIT_LIST_HEAD(&sdev->client_list);
 
-	ret = snd_sof_machine_register(sdev, plat_data);
-	if (ret < 0)
-		goto fw_trace_err;
+	/*
+	 * Register audio client.
+	 * This can fail but errors cannot be propagated.
+	 */
+	sof_client_dev_register(sdev, "sof-audio", SOF_CLIENT_AUDIO);
```

Comment:
@lyakh this is still an open for me as well. In the current design, only the core can register the clients. But this probably is a question that still needs to be addressed.

Reply:
@plbossart what about the virtio case? I havent gotten around to reading @lyakh's document yet but how would the guest register with the SOF core?

Reply:
@ranj063 Probably something exposed to qemu or the virtualization software in itself, because the guest cannot really choose anything by itself. Command line parameters of qemu and it would ask for everything needed when initializing the emulated device.

Reply:
right, it's probably more the host vhost parts that would be the SOF client. @lyakh may want to chime in here.

Reply:
My understanding of multi-client support isn't perfect, but aren't clients roughly divided along pipeline lines? I.e. as long as HDMI and SSP pipelines in the topology aren't connected, they can (and will) be separate clients, right? Whereas with virtualisation guests attach to respective host pipelines. So I don't think they can be separate clients? The only possible change might be the linking method because now I might also need a client ID, or maybe it won't be needed either, if component IDs remain unique across all clients.

Reply:
Yes, the intent was to have an SSP+DMIC card, and an HDMI card. both modeled as clients.

But indeed the virtualization is somewhat orthogonal to the notion of client, since the connections happen at the pipeline level. So you would probably have a single client for all of virtualization uses.

But that brings one open: do we expose HDMI pipelines to the guest? If yes, we'd need means for those pipelines to be supported with jack detection, and we'd need mixers on the HDMI path, which we currently don't.

one of those things where by asking a question you get more work in return :-)

Reply:
> My understanding of multi-client support isn't perfect, but aren't clients roughly divided along pipeline lines? I.e. as long as HDMI and SSP pipelines in the topology aren't connected, they can (and will) be separate clients, right? Whereas with virtualisation guests attach to respective host pipelines. So I don't think they can be separate clients? The only possible change might be the linking method because now I might also need a client ID, or maybe it won't be needed either, if component IDs remain unique across all clients.

@lyakh I have kept the component IDs unique across all client consciously. I dont currently have the notion of a client ID and besides, the FW doesnt quite partition components by clients. So creating components with the same ID across clients just caused errors in my case.

Reply:
My two cents, is it not possible/reasonable to keep a mapping on the kernel side between client component IDs and firmware component IDs? So that unique IDs can be allocated for the firmware even if they will be identical across clients.

Reply:
@paulstelian97 it is entirely possible to maintain this look-up table between the kernel and firmware. But do you think its essential? The SOF core maintains a linearly incrementing component ID that all the clients can simply query for and use it. What would be the added benefit of having the extra logic to look up the FW component IDs from the client component IDs?

Reply:
Besides conflicts when two clients request the same ID nothing really. Clients have statically selected IDs; the firmware may not need to know the specific ID the client sees but the unique, dynamically allocated ID that is selected via such a mechanism. All IPC would have this ID translated on need so the firmware isn't even made aware of multiple clients (so this would thus be a kernel-only change).

Reply:
@paulstelian97 this is a good point about conflicting comp ID requests from clients simultaneously. Adding a lock/mutex when might solve the problem

Reply:
@ranj063 So you say that the ID namespace itself should still be shared? My idea was that each client had his own namespace of IDs for components and pipelines so they don't touch the components of other clients. Or you consider that sharing might be useful? (the mapping which I mention can allow a component or pipeline to show up in multiple such namespaces, to be shared explicitly)

Reply:
@paulstelian97 in the current implementation, clients maintain a list of widgets based on the topology that is loaded when the client devices are probed. So, if they want to do perform operations on those widgets, they can look up widgets' comp ID from their own list of widgets. And this list is maintained in the private data for each client. So, it is not possible for a client to tamper with a component that belongs to another client because there would really be no way for it to look up a component from another client's widget list. Does that make sense?

Reply:
@ranj063 I think so, so the clients know the actual component IDs (without namespacing) but can't work with other components that don't belong to them. Then there must be some dynamic allocation of component IDs (topology doesn't hardcode the actual IDs used at runtime)? Otherwise we will see clashes. I haven't fully grasped the topology loading code so if this is there just point it out and I will be okay with that.

Reply:
@paulstelian97 here's where the comp ID's are assigned while parsing topology. The next_comp_id field of sdev is a linearly increasing value with each widget that is created in the DSP FW. And this is maintained by the core. So lets say you have 2 clients, one for I2S pipelines with N widgets and one for HDMI pipelines with M widgets. The valid comp ID's for the I2S client would run from 0 to N-1 and the valid comp ID's for the HDMI client would run from N -> N + M - 1. And so typically, when a client wants to perform operations on its components, it would look it up from its own widget list and the components in the other clients' widget list would be off limits.
https://github.com/thesofproject/linux/blob/4cef169b7bf4c6033e8e478921349ae23e5ab300/sound/soc/sof/topology.c#L2217

Reply:
Oh great, then my issue is invalid :) Now my only thing is if there are components that may be explicitly shared between clients. Is this not supported or? (again my understanding is pretty limited, just curious). You would bridge the connections between clients using custom port components or there is some sharing?

Reply:
@paulstelian97 as I understand, if there are shared components between pipelines, it is better to keep them in the same client thereby meaning that they also remain part of the same card. And this is the reason why we are opting to keep the DMIC pipelines clubbed together with the codec pipelines so that when there is a need to add features such as echo cancellation, we are not dealing with multiple cards.